src/common/admin_socket.h

```diff
@@ -22,19 +22,34 @@
 #include <thread>
 
 #include "include/buffer.h"
+#include "common/ref.h"
 #include "common/cmdparse.h"
 
 class AdminSocket;
 class CephContext;
+class MCommand;
 
 using namespace std::literals;
 
 inline constexpr auto CEPH_ADMIN_SOCK_VERSION = "2"sv;
 
 class AdminSocketHook {
 public:
-  virtual bool call(std::string_view command, const cmdmap_t& cmdmap,
-		    std::string_view format, ceph::buffer::list& out) = 0;
+  // NOTE: the sync handler doesn't take an input buffer currently because
+  // no users need it yet.
+  virtual int call(std::string_view command, const cmdmap_t& cmdmap,
+		   std::string_view format, ceph::buffer::list& out) = 0;
+  virtual void call_async(
+    std::string_view command,
+    const cmdmap_t& cmdmap,
+    std::string_view format,
+    const bufferlist& inbl,
+    std::function<void(int,const std::string&,bufferlist&)> on_finish) {
+    // by default, call the synchronous handler and then finish
+    bufferlist out;
+    int r = call(command, cmdmap, format, out);
```

Comment:
This would seem like a good opportunity to standardize on JSON (as discussed in CLT) and transition to some standard output format. We could have a standard return code (as an int), error message, ceph version, etc. Even if it's not trivial to convert all commands to take a `JSONFormatter`, we could try to decode the bufferlist as json and then stick it in some output:

`{rc=0,errmsg=null,output={"custom":"output"},version="ceph version..."}`

otherwise if it doesn't decode:

`{rc=0,errmsg=null,output="output string that is not JSON",version="ceph version..."}`

What do you think?

It'd also be interesting to consider that some commands may run asynchronously and we could eventually have a way to query running commands and wait to get output.

Reply:
I could see us going so far as to pass a Formatter* to these call methods.  My main reservation is that JSON (at least) isn't fully expressive, so if we ever want to return binary data (say, a binary dump of the osdmap), we can't.

I'm inclined to go command-by-command and make sure what we have now is dumping JSON wherever possible?

Reply:
Sorry didn't see the notification for your response.

> I could see us going so far as to pass a Formatter* to these call methods. My main reservation is that JSON (at least) isn't fully expressive, so if we ever want to return binary data (say, a binary dump of the osdmap), we can't.

Agreed. Formatter makes more sense in that case.

> I'm inclined to go command-by-command and make sure what we have now is dumping JSON wherever possible?

I'd also advocate for a better data description language, maybe Lua? I could whip together a formatter for Lua if there's interest.